docs/client-conventions.md

```diff
@@ -0,0 +1,164 @@
+
+## Associate modifications with Revisions
+
+Every time the client changes a Service or Configuration in a way that results
+in a new Revision, it should change the label `client.knative.dev/nonce` to a
+new random value, chosen from a wide enough space to avoid collisions.
+
+This way, the client can make a label selector query for Revisions with this
+nonce value to find the Revision the particular change generated. The client can
+use that revision to, for example, inform the user about the readiness of their
+requested change.
+
+### In the presence of non-conventional clients
+
+If an unconventional client neglects to reset this value, the label selector
+query could result in more than one revision; the client should pick the one
+with the lowest `configurationGeneration` in this case.
+
+## Force creation of a new Revision
+
+The way to deploy new code with a previously-used tag is to make a new Revision,
+which the Revision controller will re-pull and lock it to the current image at
+that tag. Unfortunately Knative won't create a new Revision if you don't change
+the Configuration.
+
+The same `client.knative.dev/nonce` annotation can help in forcing the creation
+of a new Revision; if the nonce is changed, the Configuration controller must
+make one even if nothing else has changed.
+
+## Change non-code attributes
+
+When the user specifies they'd like to change an environment variable (or a
+memory allocation, or a concurrency setting...), and does not specify that
+they'd like to deploy a change in code, the user would be quite surprized to
+find the newest image at their deployed tag running in the cloud.
+
+### General idea
+
+Since the Revision controller will resolve an image tag for every Revision
+creation, we need a way to express a non-code change. Clients should do this by
+changing the `image` field to be a digest-based image URL supplied by the
+Revision `status.imageDigest` field, while marking the original tag-based user
+intent in an annotation.
+
+### Procedure
+
+ 1. Get the current state of the Service in question.
+ 2. Get a **base revision**, the Revision corresponding to the fetched state of
+    the Service: Issue a label selector query for revisions with the
+    `client.knative.dev/nonce` label listed in the
+    `revisionTemplate.metadata`. If this yields exactly one Revision (after
+    waiting a sensible timeout for a revision to appear with that nonce), that
+    is the base revision. If not, fetch the `latestCreatedRevision` from the
+    status, and uses that as the base revision.
+ 3. Copy the `status.imageDigest` field from the base revison into the `image`
+    field of the Service. This ensures the running code stays the same.
+ 4. Make whatever other modifications to the Service.
+ 5. Add the `client.knative.dev/user-image` annotation to the Service,
+    containing the original tag-based URL of the image.
+ 6. Add the nonce label, as usual.
+ 7. Post the resulting Service to create a new Revision.
```

Comment:
You don't really talk about the "restoration" process for `user-image`, which is I think one of the biggest gotchas of non-conventional clients, e.g. they just keep redeploying the same code (meanwhile, a conventional UI thinks that's what the user intended).